src/boot/memory.c

```diff
@@ -115,139 +110,147 @@ void move_segment_table_to_dmem(void) {
+    header->magic = MAIN_POOL_FREEABLE_HEADER_MAGIC_RIGHT;
+    header->id = id;
+    header->ptr = regionEnd;
+
+    region->end -= sizeWithHeader;
+
+    return header->data;
+}
+
+static void* main_pool_region_try_alloc_from_end_aligned_freeable(struct MainPoolRegion* region, u8 id, u32 size, u32 alignment) {
+    u8* region_end = region->end;
+    u8* new_end = (u8*) DOWN(region_end - size, alignment) - sizeof(struct MainPoolFreeableHeader);
+    if (new_end < region->start)
+        return NULL;
+
+    size = region_end - new_end;
+
+    struct MainPoolFreeableHeader* header = (struct MainPoolFreeableHeader*) new_end;
+    header->magic = MAIN_POOL_FREEABLE_HEADER_MAGIC_RIGHT;
+    header->id = id;
+    header->ptr = region_end;
+
+    region->end -= size;
+
+    return header->data;
+}
+
+void *main_pool_alloc_aligned(u32 size, u32 alignment) {
+    if (!alignment)
+        alignment = 16;
+
+    size = ALIGN4(size);
+    return main_pool_region_try_alloc_from_start_aligned(&gMainPool.regions[0], size, alignment);
+}
+
+void *main_pool_alloc_freeable(u32 size) {
+    size = ALIGN4(size) + sizeof(struct MainPoolFreeableHeader);
+    return main_pool_region_try_alloc_from_end_freeable(&gMainPool.regions[0], 0, size);
+}
+
+void *main_pool_alloc_aligned_freeable(u32 size, u32 alignment) {
+    if (!alignment)
+        alignment = 16;
+
+    size = ALIGN4(size);
+    return main_pool_region_try_alloc_from_end_aligned_freeable(&gMainPool.regions[0], 0, size, alignment);
 }
 
 /**
  * Free a block of memory that was allocated from the pool. The block must be
  * the most recently allocated block from its end of the pool, otherwise all
  * newer blocks are freed as well.
- * Return the amount of free space left in the pool.
  */
-u32 main_pool_free(void *addr) {
-    struct MainPoolBlock *block = (struct MainPoolBlock *) ((u8 *) addr - 16);
-    struct MainPoolBlock *oldListHead = (struct MainPoolBlock *) ((u8 *) addr - 16);
-
-    if (oldListHead < sPoolListHeadL) {
-        while (oldListHead->next != NULL) {
-            oldListHead = oldListHead->next;
-        }
-        sPoolListHeadL = block;
-        sPoolListHeadL->next = NULL;
-        sPoolFreeSpace += (uintptr_t) oldListHead - (uintptr_t) sPoolListHeadL;
+void main_pool_free(void *addr) {
+    const struct MainPoolFreeableHeader* header = (struct MainPoolFreeableHeader*) ((u8*) addr - sizeof(struct MainPoolFreeableHeader));
+    if (header->magic == MAIN_POOL_FREEABLE_HEADER_MAGIC_RIGHT) {
+        struct MainPoolRegion* region = &gMainPool.regions[header->id];
+        region->end = header->ptr;
     } else {
-        while (oldListHead->prev != NULL) {
-            oldListHead = oldListHead->prev;
-        }
-        sPoolListHeadR = block->next;
-        sPoolListHeadR->prev = NULL;
-        sPoolFreeSpace += (uintptr_t) sPoolListHeadR - (uintptr_t) oldListHead;
+        DEBUG_ASSERT("Incorrect magic for free");
```

Comment:
This doesn't do anything, use `assert(FALSE, message);` for errors that execute with debug only (tbh I didn't even know this macro was in the repo, it should probably be removed...)

Reply:
Actually this should probably be a non-debug assertion, in which case use `aggress(header->magic == MAIN_POOL_FREEABLE_HEADER_MAGIC_RIGHT, "Incorrect magic for free");` and scrap the if statement.

Reply:
Applied suggested improvement.